internal/layer/read.go

```diff
@@ -0,0 +1,326 @@
+package layer
+
+import (
+	"context"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	cdl "github.com/mimiro-io/common-datalayer"
+	egdm "github.com/mimiro-io/entity-graph-data-model"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func (d *Dataset) Changes(since string, limit int, latestOnly bool) (cdl.EntityIterator, cdl.LayerError) {
+	if latestOnly {
+		// the layer does not know if the given table is a "change" table or not, so we cannot support this mode with confidence
+		return nil, cdl.Err(fmt.Errorf("latest only operation not supported"), cdl.LayerNotSupported)
+	}
+
+	mapper := cdl.NewMapper(d.logger, d.datasetDefinition.IncomingMappingConfig, d.datasetDefinition.OutgoingMappingConfig)
+	iter, err := d.newIterator(mapper, since, limit)
+	if err != nil {
+		return nil, err
+	}
+
+	return iter, nil
+}
+
+func (d *Dataset) Entities(from string, limit int) (cdl.EntityIterator, cdl.LayerError) {
+	// the layer does not know if the given table is a "change" table or not, so implement /entities as /changes
+	// TODO: consider adding source config options to allow for different behavior
+	return d.Changes(from, limit, false)
+}
+
+func getConfigProperty(config map[string]interface{}, key string) string {
+	val, ok := config[key]
+	if !ok {
+		return ""
+	}
+	valStr, ok := val.(string)
+	if !ok {
+		return ""
+	}
+	return valStr
+}
+
+func (d *Dataset) newIterator(mapper *cdl.Mapper, since string, limit int) (*dbIterator, cdl.LayerError) {
+	entityColumn := getConfigProperty(d.datasetDefinition.SourceConfig, EntityColumn)
+	sinceCol := getConfigProperty(d.datasetDefinition.SourceConfig, SinceColumn)
+	ctx := context.Background() // no timeout because we want to support long running stream operations
+
+	db := d.db.db
+
+	var maxSince sql.NullTime
+	var nextToken string
+	if sinceCol != "" {
+		// build max since query
+		maxSinceQuery := "SELECT MAX(" + sinceCol + ") AS \"_MAX_SINCE\" FROM " + d.datasetDefinition.SourceConfig[TableName].(string)
+		rows, err := db.QueryContext(ctx, maxSinceQuery)
+		if err != nil {
+			return nil, cdl.Err(err, cdl.LayerErrorInternal)
+		}
+		defer func() {
+			rows.Close()
+		}()
+
+		hasData := rows.Next()
+		if !hasData {
+			d.logger.Error("failed to get max since", "error", "no data")
+			return nil, cdl.Err(fmt.Errorf("failed to get max since"), cdl.LayerErrorInternal)
+		}
+
+		err = rows.Scan(&maxSince)
+		if err != nil {
+			d.logger.Error("failed to scan max since", "error", err)
+			return nil, ErrQuery(err)
+		}
+
+		newSince := maxSince.Time.Format("2006-01-02 15:04:05.000000")
```

Comment:
what if the since column is a different datatype? it will probably return an erro in line 78, but it might not be clear to users that the column must be a Time

Reply:
Improved the error message to give guidance on correct type.